src/ptr.rs

```diff
@@ -0,0 +1,331 @@
+        matches!(self.tag, ExprTag::Nil)
+        // FIXME: check value also, probably
+    }
+    pub const fn is_cons(&self) -> bool {
+        matches!(self.tag, ExprTag::Cons)
+    }
+
+    pub const fn is_atom(&self) -> bool {
+        !self.is_cons()
+    }
+
+    pub const fn is_list(&self) -> bool {
+        matches!(self.tag, ExprTag::Nil | ExprTag::Cons)
+    }
+
+    pub const fn is_opaque(&self) -> bool {
+        self.raw.is_opaque()
+    }
+
+    pub const fn as_cons(self) -> Option<Self> {
+        if self.is_cons() {
+            Some(self)
+        } else {
+            None
+        }
+    }
+
+    pub const fn as_list(self) -> Option<Self> {
+        if self.is_list() {
+            Some(self)
+        } else {
+            None
+        }
+    }
+
+    pub fn index(tag: ExprTag, idx: usize) -> Self {
+        Ptr {
+            tag,
+            raw: RawPtr::Index(idx),
+            _f: Default::default(),
+        }
+    }
+
+    pub fn opaque(tag: ExprTag, idx: usize) -> Self {
+        Ptr {
+            tag,
+            raw: RawPtr::Opaque(idx),
+            _f: Default::default(),
+        }
+    }
+
+    pub fn null(tag: ExprTag) -> Self {
+        Ptr {
+            tag,
+            raw: RawPtr::Null,
+            _f: Default::default(),
+        }
+    }
+}
+
+impl<F: LurkField> From<char> for Ptr<F> {
+    fn from(c: char) -> Self {
+        Self {
+            tag: ExprTag::Char,
+            raw: RawPtr::Index(u32::from(c) as usize),
+            _f: Default::default(),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct ContPtr<F: LurkField> {
+    pub tag: ContTag,
+    pub raw: RawPtr,
+    pub _f: PhantomData<F>,
+}
+
+#[allow(clippy::derived_hash_with_manual_eq)]
+impl<F: LurkField> Hash for ContPtr<F> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.tag.hash(state);
+        self.raw.hash(state);
+    }
+}
+
+impl<F: LurkField> ContPtr<F> {
+    pub fn new(tag: ContTag, raw: RawPtr) -> Self {
+        Self {
+            tag,
+            raw,
+            _f: Default::default(),
+        }
+    }
+    pub const fn is_error(&self) -> bool {
+        matches!(self.tag, ContTag::Error)
+    }
+
+    pub fn index(tag: ContTag, idx: usize) -> Self {
+        ContPtr {
+            tag,
+            raw: RawPtr::Index(idx),
+            _f: Default::default(),
+        }
+    }
+
+    pub fn opaque(tag: ContTag, idx: usize) -> Self {
+        ContPtr {
+            tag,
+            raw: RawPtr::Index(idx),
+            _f: Default::default(),
+        }
+    }
+
+    pub fn null(tag: ContTag) -> Self {
+        ContPtr {
+            tag,
+            raw: RawPtr::Null,
+            _f: Default::default(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Arbitrary))]
+// Note: the trait bound E: Tag is not necessary in the struct, but it makes the proptest strategy more efficient.
```

Comment:
This comment is something you might link to in your upcoming issue @johnchandlerburnham